### ISSUE TITLE
Illegal ambiguous match with flags

### ISSUE THREAD

**Issue by jjwon0** [bazelbuild/bazel#17697]
### Description of the bug:

When combining a flag with another `config_setting`, Bazel seems to think that one is not a more specialized version of the other. However, I can't understand why that is.

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

I have a simple repro case here: https://github.com/jjwon0/bazel-select-mre/

### Which operating system are you running Bazel on?

linux

### What is the output of `bazel info release`?

release 6.1.0

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

n/a

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?

```text
n/a
```

### Have you found anything relevant by searching the web?

I read through all the recent issues related to `select()` but nothing seemed to be an exact match. I also read the configuration docs through a few times.

### Any other information, logs, or outputs that you want to share?

It's totally possible that I don't understand what it means to be specialized, but I would have thought that matching one of many attributes in a `config_setting_group` would do the trick.

This also doesn't seem to happen on Bazel 5.1.1.

**Comment by jjwon0** [comment]
Some other relevant details we found:
- The difference seems to be [this new line](https://github.com/bazelbuild/bazel/blob/6.1.0/src/main/java/com/google/devtools/build/lib/analysis/config/ConfigMatchingProvider.java#L89) which tries to discern the specialization (before, that additional condition [didn't exist](https://github.com/bazelbuild/bazel/blob/5.1.0/src/main/java/com/google/devtools/build/lib/analysis/config/ConfigMatchingProvider.java#L84)).
- For resolving matches of a `config_setting` with `config_setting_group`, it seems to have some dependency on the ordering of the `match_all` field. I've added repro for this on [another branch](https://github.com/jjwon0/bazel-select-mre/tree/array-ordering) since I'm not really sure if this is the same thing or something different. Whatever is going on, it certainly doesn't seem to be doing what I would think, which is combining the values/constraints in the `config_setting_group` and comparing that against the `config_setting`.

**Comment by jjwon0** [comment]
Actually just as I typed that up, I seem to be able to reproduce the ordering issue without adding more settings. Now, if I swap `:_foo` and `:linux` in the example on [`main`](https://github.com/jjwon0/bazel-select-mre/) it results in the failure, but succeeds as-is. Is there some local state that could be causing issues here? I tried `bazel shutdown` and `bazel clean`; neither seems to affect it.

**Comment by aranguyen** [comment]
@jjwon0 

> Now, if I swap :_foo and :linux in the example on [main](https://github.com/jjwon0/bazel-select-mre/) it results in the failure, but succeeds as-is

Do you mean, it matters if we have 

```
selects.config_setting_group(
    name = "foo",
    match_all = [
        ":linux",
        ":_foo",
    ],
)
```

vs 

```
selects.config_setting_group(
    name = "foo",
    match_all = [
        ":_foo",
        ":linux",
    ],
)
```

And the later one succeeded without the specialization error?

**Comment by jjwon0** [comment]
Yes, that's right. I noticed that for some reason, order seemed to matter. However it wasn't consistent between the separate times I tried to reproduce it so it feels like there is some caching bug.

**Comment by JohanLaineTobii** [comment]
Hello! I'm running into the same issue but from what I can tell no ordering makes it work. Similar example as above, but directly with `constraint_value`s from `@platforms`:
```
load("@bazel_skylib//lib:selects.bzl", "selects")

selects.config_setting_group(
    name = "linux_x86_64",
    match_all = [
        "@platforms//os:linux",
        "@platforms//cpu:x86_64",
    ],
)

filegroup(
    name = "example",
    srcs = select({
        "@platforms//os:linux": ["file1.txt"],
        ":linux_x86_64": ["file2.txt"],
    }),
)
```

It surprises me a bit that this doesn't work, I'd assume this construction to be the most common way of achieving specialization in select statements.
